contribs/gmf/src/directives/layertree.js

```diff
@@ -0,0 +1,296 @@
+ * and a ngeo.Layertree. The default used template can be overrided by setting
+ * the constant 'gmf.layertreeTemplateUrl'. The controller used by this
+ * directive defines some fonctions for each node that are created by a default
+ * template. This default template can be overrided by setting the constant
+ * 'gmf.layertreeTemplateUrl'.
+ *
+ * @example
+ * <gmf-layertree
+ *   gmf-layertree-themes="ctrl.themes"
+ *   gmf-layertree-map="ctrl.map"
+ *   gmf-layertree-wmsurl="ctrl.wmsUrl">
+ * </gmf-layertree>
+ *
+ * @return {angular.Directive} The directive specs.
+ * @ngInject
+ * @ngdoc directive
+ * @ngname gmfLayertreeDirective
+ */
+
+gmf.layertreeDirective = function() {
+  return {
+    scope: {
+      'getMapFn': '&gmfLayertreeMap',
+      'getWmsUrlFn': '&gmfLayertreeWmsurl',
+      'themes': '=gmfLayertreeThemes'
+    },
+    controller: 'GmfLayertreeController',
+    controllerAs: 'gmfLayertreeCtrl',
+    template:
+        '<div ngeo-layertree="gmfLayertreeCtrl.tree" ' +
+        'ngeo-layertree-map="gmfLayertreeCtrl.map" ' +
+        'ngeo-layertree-nodelayer="gmfLayertreeCtrl.getLayer(node)" ' +
+        'ngeo-layertree-templateurl="' + gmf.layertreeTemplateUrl + '"' +
+        '<div>'
+  };
+};
+
+gmfModule.directive('gmfLayertree', gmf.layertreeDirective);
+
+
+
+/**
+ * @param {angular.Scope} $scope The directive's scope.
+ * @param {ngeo.LayerManager} ngeoLayerManager Ngeo Layer Manager.
+ * @constructor
+ * @export
+ * @ngInject
+ * @ngdoc controller
+ * @ngname gmfLayertreeController
+ */
+gmf.LayertreeController = function($scope, ngeoLayerManager) {
+
+  var map = $scope['getMapFn']();
+  goog.asserts.assertInstanceof(map, ol.Map);
+
+  /**
+   * @type {!ol.Map}
+   * @export
+   */
+  this.map = map;
+
+  var wmsUrl = $scope['getWmsUrlFn']();
+  goog.asserts.assert(goog.isString(wmsUrl));
+
+  /**
+   * @type {string}
+   * @private
+   */
+  this.wmsUrl_ = wmsUrl;
+
+  /**
+   * @type {ngeo.LayerManager}
+   * @private
+   */
+  this.layerManager_ = ngeoLayerManager;
+  ngeoLayerManager.init(this.map);
+
+  /**
+   * @type {Object.<string, function((boolean|undefined)):(boolean|undefined)>}
+   * @private
+   */
+  this.checkboxGettersSetters_ = {};
+
+  /**
+   * @type {Object.<string, function():(string|undefined)>}
+   * @private
+   */
+  this.styleStates_ = {};
+
+  //FIXME Choose a theme in gmf datasource
+  $scope.$watch('themes', goog.bind(function(newVal) {
+    if (goog.isDef(newVal)) {
+      this['tree'] = newVal[3];
+    }
+  }, this));
+};
+
+
+/**
+ * Create a return a layer corresponding to the ngeo.layerTree's node.
+ * Can currently only creates WMS layers (internal or external) and WMTS layers.
+ * @param {GmfThemesNode} node Layer tree node.
+ * @return {ol.layer.Layer} The OpenLayers layer or group for the node.
+ * @export
+ */
+gmf.LayertreeController.prototype.getLayer = function(node) {
+  var wmsUrl = this.wmsUrl_;
+  var layerMgr = this.layerManager_;
+
+  var makeLayers = function(node) {
```

Comment:
Because I call itself in the function (recursivity).
If the node 'is' a layer group. I want that this node 'contains' a layergroup with all its sublayers or/and subgroups.
Perhaps I should create an outer private function instead of this inner function ?